lib/auth/auth_dtls.c

```diff
@@ -0,0 +1,959 @@
+ * time (wall clock time) isn't necessary, just the relative time in milliseconds.
+ *
+ * @param val     Timer value
+ * @param reset   If non-zero then set current milliseconds into the time var.
+ *
+ * @return 0 if resetting the time, else the time difference in milliseconds.
+ */
+static unsigned long auth_tls_timing_get_timer(struct mbedtls_timing_hr_time *val, int reset)
+{
+	int64_t delta;
+	int64_t *mssec = (int64_t *) val;
+	int64_t cur_msg = k_uptime_get();
+
+	if (reset) {
+		*mssec = cur_msg;
+		return 0;
+	}
+
+	delta = cur_msg - *mssec;
+
+	return (unsigned long )delta;
+}
+
+/**
+ * Set delays to watch, final and intermediate delays.
+ *
+ * @param data    Timing delay context.
+ * @param int_ms  Intermediate delay in milliseconds.
+ * @param fin_ms  Final delay in milliseconds.
+ *
+ */
+static void auth_tls_timing_set_delay(void *data, uint32_t int_ms, uint32_t fin_ms)
+{
+	mbedtls_timing_delay_context *ctx = (mbedtls_timing_delay_context *) data;
+
+	ctx->int_ms = int_ms;
+	ctx->fin_ms = fin_ms;
+
+	if (fin_ms != 0) {
+		(void) auth_tls_timing_get_timer(&ctx->timer, 1);
+	}
+}
+
+/**
+ *  Determine the delay result.
+ *
+ *  @param data   Pointer to delay context.
+ *
+ *  @return -1 if cancelled, 0 if none of the delays is expired,
+ *           1 if the intermediate delay only is expired,
+ *           2 if the final delay is expired
+ */
+static int auth_tls_timing_get_delay(void *data)
+{
+	mbedtls_timing_delay_context *ctx = (mbedtls_timing_delay_context *)data;
+	int64_t elapsed_ms;
+
+	if (ctx->fin_ms == 0) {
+		return -1;
+	}
+
+	elapsed_ms = auth_tls_timing_get_timer(&ctx->timer, 0);
+
+	if (elapsed_ms >= (int64_t)ctx->fin_ms) {
+		return 2;
+	}
+
+	if (elapsed_ms >= (int64_t)ctx->int_ms) {
+		return 1;
+	}
+
+	return 0;
+}
+
+
+/**
+ * Function called by Mbed stack to print debug messages.
+ *
+ * @param ctx     Context
+ * @param level   Debug level
+ * @param file    Source filename of debug log entry.
+ * @param line    Line number of debug log entry.
+ * @param str     Debug/Log message.
+ */
+static void auth_mbed_debug(void *ctx, int level, const char *file,
+			    int line, const char *str)
+{
+	const char *p, *basename;
+
+	/**
+	 * @brief   Need to define const string here vs. const char *fmt = "[%s:%d] %s"
+	 *          because the LOG_ERR(), LOG_* macros can't handle a pointer.
+	 */
+#define LOG_FMT  "[%s:%d] %s"
+
+	(void)ctx;
+
+	if (!file || !str) {
+		return;
+	}
+
+	/* Extract basename from file */
+	for (p = basename = file; *p != '\0'; p++) {
+		if (*p == '/' || *p == '\\') {
+			basename = p + 1;
+		}
+	}
+
+
+	switch (level) {
+
+	case 0:
+	{
+		LOG_ERR(LOG_FMT, log_strdup(basename), line, log_strdup(str));
+		break;
+	}
+
+	case 1:
+	{
+		LOG_WRN(LOG_FMT, log_strdup(basename), line, log_strdup(str));
+		break;
+	}
+
+	case 2:
+	{
+		LOG_INF(LOG_FMT,  log_strdup(basename), line, log_strdup(str));
+		break;
+	}
+
+	case 3:
```

Comment:
use `__fallthrough` attribute in these cases.

Reply:
Fixed.